apps/lead/serializers.py

```diff
@@ -717,3 +720,51 @@ def validate(self, data):
         data['project'] = self.project
         data['user'] = self.current_user
         return data
+
+
+class DeduplicationCallbackSerializer(serializers.Serializer):
+    client_id = serializers.CharField()
+    lead_id = serializers.IntegerField()
+    duplicate_lead_ids = serializers.ListField(child=serializers.IntegerField())
+
+    def __init__(self, *args, **kwargs):
+        self.lead = None
+        self.duplicate_leads = None
+        super().__init__(*args, **kwargs)
+
+    def validate_lead_id(self, lead_id):
+        self.lead = Lead.objects.filter(pk=lead_id).first()
+        if not self.lead:
+            raise serializers.ValidationError("Lead does not exist")
+        return lead_id
+
+    def validate_duplicate_lead_ids(self, lead_ids):
+        if not lead_ids:
+            raise serializers.ValidationError("Empty lead ids")
+        self.duplicate_leads = Lead.objects.filter(pk__in=lead_ids)
+        if self.duplicate_leads.count() != len(lead_ids):
+            raise serializers.ValidationError("All of the duplicate lead ids do not exist")
+        return lead_ids
+
+    def validate(self, data):
+        if self.lead is None or self.duplicate_leads is None:
+            raise serializers.ValidationError("Invalid data. Provide lead_id and duplicate_lead_ids")
+
+        # Also validate if lead_id corresponds to the client id
+        try:
+            lead = LeadExtraction.get_lead_from_client_id(data['client_id'])
+        except LeadExtraction.Exception.InvalidTokenValue:
+            raise serializers.ValidationError("Invalid token")
+        if lead.id != data['lead_id']:
+            raise serializers.ValidationError("Inconsistent lead_id and client_id")
```

Comment:
Why error message has `client_id`?

Reply:
Because it is checking whehter the lead  id is associated with corresponding client_id or not. Such checks are not done in other callbacks but it's done here.